### PR TITLE
 [Hardware][Ascend] Adapt Qwen3 TTS for 310P

### PR DESCRIPTION
**Pull request by zyz111222**
## Purpose

  This PR adapts Qwen3-TTS for Ascend 310P execution.
Related to vllm-project/vllm-ascend#9712; this PR adds the vLLM-Omni-side Qwen3-TTS adaptation for Ascend 310P.

  The change keeps the generic NPU path unchanged where possible and gates 310P-specific behavior behind Ascend 310P detection. It covers the Qwen3-TTS code predictor attention path, reference-audio frontend runtime, NPU worker/model runner routing, and NPU dependency alignment.

## Test Plan
Run focused CPU unit tests for the new 310P helper.
Run existing Qwen3-TTS code predictor dtype tests with the new 310P import mocked.
Run git whitespace checks against `origin/main`.

## Test Result

seed-tts-eval results: 

image

performance benchmark:
image
Using a sample from seed-tts-eval as an example to test RTF, the sample corresponds to 25 text characters, the reference audio duration is 4.7s, and the RTF result is 1.61.

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [x] The test plan. Please provide the test scripts & test commands. Please state the reasons if your codes don't require additional test scripts. For test file guidelines, please check the [test style doc](https://docs.vllm.ai/projects/vllm-omni/en/latest/contributing/ci/tests_style/)
- [x] The test results. Please paste the results comparison before and after, or the e2e results.
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model. **Please run `mkdocs serve` to sync the documentation editions to `./docs`.**
- [ ] (Optional) Release notes update. If your change is user-facing, please update the release notes draft.
</details>

**BEFORE SUBMITTING,** please read <https://github.com/vllm-project/vllm-omni/blob/main/CONTRIBUTING.md>
(anything written below this line will be removed by GitHub Actions)

### CODE REVIEW

File: vllm_omni/model_executor/models/common/qwen3_code_predictor.py [hunk 1]
```diff
@@ -69,6 +70,28 @@ def _rotate_half(x: torch.Tensor) -> torch.Tensor:
     return torch.cat((-x2, x1), dim=-1)
 
 
+def _align_up(value: int, alignment: int) -> int:
```
Comment (gcanlin) [inline vllm_omni/model_executor/models/common/qwen3_code_predictor.py]: Let's move these code that's only for 310p to `platforms/npu/310p/` and use patch mechanism to enable it.

File: vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py [hunk 2]
```diff
@@ -38,6 +38,11 @@
 from vllm.logger import init_logger
 from vllm.multimodal.audio import AudioResampler
 
+from vllm_omni.platforms.npu.ascend_310p import (
```
Comment (gcanlin) [inline vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py]: ditto

File: vllm_omni/platforms/npu/worker/base.py [hunk 3]
```diff
@@ -31,6 +37,16 @@ def __init__(self, *args, **kwargs):
                 local_rank=self.local_rank,
             )
 
+    def _init_device(self):
+        device = super()._init_device()
+        if is_310p():
+            from vllm_omni.platforms.npu.ascend_310p import disable_jit_compile
+
+            # 310P uses the vLLM-Ascend 310 worker, then disables torch_npu JIT
+            # for the Qwen3-TTS serving path after parent device initialization.
```
Comment (gcanlin) [inline vllm_omni/platforms/npu/worker/base.py]: Didn't we do it in vLLM Ascend? Here we're reusing the worker from upstream. Seems that we should have it natually.

Comment (zyz111222) [reply vllm_omni/platforms/npu/worker/base.py]: `NPUARWorker` and `NPUGenerationWorker` override `init_device()`; since they both inherit from `OmniWorkerBase`, the configuration for the 310 chip is overridden here. Comments have been updated to reflect this.

**Review by hsliuustc0106 — COMMENTED** [review]
Hardware adaptation for Ascend 310P looks appropriate. Gates 310P-specific behavior behind hardware checks, keeping the generic NPU path unchanged.

File: vllm_omni/model_executor/models/common/qwen3_code_predictor.py (hunk 1 above)
Comment (zyz111222) [reply vllm_omni/model_executor/models/common/qwen3_code_predictor.py]: I have refactored the intrusive modifications for 310P; the current version does not affect the functionality of the main branch.

File: vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py (hunk 2 above)
Comment (zyz111222) [reply vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py]: I have refactored the intrusive modifications for 310P; the current version does not affect the functionality of the main branch.

**Review by gcanlin — APPROVED** [review]
LGTM now.

**Comment by gcanlin** [comment]
Can you also run a benchmark for performance?

File: vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py [hunk 4]
```diff
@@ -329,6 +329,7 @@ def __init__(
         self._tts_pad_embed_buffer = tts_pad_embed
         self._encode_ref_audio_batch_fn = encode_ref_audio_batch
         self._speaker_cache = speaker_cache
+        self._embedding_dtype = torch.bfloat16
```
Comment (hsliuustc0106) [inline vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py:332]: why we make this change? it does not make sense to me

Comment (gcanlin) [reply vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py:332]: Abstracting it to one field so that NPU 310p can patch it once because 310p only supports float16.

Comment (hsliuustc0106) [reply vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py:332]: but is fixed in init, we can not make changes anymore

Comment (gcanlin) [reply vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py:332]: Yes, we expect to try refactoring it to make it able to be changed by CLI arguments. But currently, we make it step by step. Abstracting it as the first step. Then, we only need to patch `__init__` instead of every layer.

Comment (zyz111222) [reply vllm_omni/model_executor/models/qwen3_tts/prompt_embeds_builder.py:332]: In the original code, bf16 was written directly in the code line. To make patching easier, we simply extracted this value.

**Comment by Michael-24** [comment]
大概什么时候合pr呢🥹

**Comment by zyz111222** [comment]
> 大概什么时候合pr呢🥹

尽量这两天

**Comment by Sky6662025** [comment]
Atlas300IDUO 96G 这个卡能通过vllm-omni   运行Qwen/Qwen-Image-2512模型？
